dsl_parser/functions.py

```diff
@@ -772,115 +734,94 @@ def parse(raw_function, scope=None, context=None, path=None):
     return raw_function
 
 
-def evaluate_functions(payload, context,
-                       get_node_instances_method,
-                       get_node_instance_method,
-                       get_node_method,
-                       get_secret_method,
-                       get_capability_method):
+def evaluate_node_functions(node, storage):
+    handler = runtime_evaluation_handler(storage)
+    scan.scan_node_template(node, handler, replace=True)
+    return node
+
+
+def evaluate_node_instance_functions(instance, storage):
+    handler = runtime_evaluation_handler(storage)
+    scan.scan_properties(
+        instance.get('runtime_properties', {}),
+        handler,
+        scope=scan.NODE_TEMPLATE_SCOPE,
+        context=instance,
+        path='{0}.runtime_properties'.format(
+            instance['id']),
+        replace=True)
+    return instance
+
+
+def evaluate_functions(payload, context, storage):
     """Evaluate functions in payload.
 
     :param payload: The payload to evaluate.
     :param context: Context used during evaluation.
-    :param get_node_instances_method: A method for getting node instances.
-    :param get_node_instance_method: A method for getting a node instance.
-    :param get_node_method: A method for getting a node.
-    :param get_secret_method: A method for getting a secret.
-    :param get_capability_method: A method for getting a capability.
+    :param storage: Storage backend for runtime function evaluation
     :return: payload.
     """
-    handler = runtime_evaluation_handler(get_node_instances_method,
-                                         get_node_instance_method,
-                                         get_node_method,
-                                         get_secret_method,
-                                         get_capability_method)
+    scope = None
+    if 'source' in context and 'target' in context:
+        scope = scan.NODE_TEMPLATE_RELATIONSHIP_SCOPE
+    elif 'self' in context:
```

Comment:
Can this ever fall through?

Reply:
i dont think so

in general this is ugly, and somewhat related to CY-1676 (cleaning up get-property)
this is tightly coupled to https://github.com/cloudify-cosmo/cloudify-common/blob/00f0abbf9f3d0168fe2aa61f44870965ad68ca88/cloudify/endpoint.py#L135-L142

an alternative would be to probably have an explicit `type` field, or have that other piece of code call a more specific thing than just the generic "evaluate functions" endpoint

Reply:
me too but that's what CY-1676 is about, not this issue :P